# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/swgl/src/vector_type.h
# Commit: ff83cdd3dd3d
# Full Hash: ff83cdd3dd3dce67f53d23bddc00c6253730e496
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2021-02-17 09:45:59
# Regressor Bug: 1692731
# File Overlap Count: 1
# Description:
#   Bug 1692731 - Accelerate YUV 422 compositing in SWGL. r=jrmuizel
#   
#   For YUV 422 video, when we are sampling UV planes at half the resolution of the
#   Y plane, we can interpolate from 2 samples for the UV planes as an approximation
#   of the 4 samples, allowing us to better pack the math into SIMD vectors and
# ==============================================================================

diff -r 1b5b53edd1d4 -r ff83cdd3dd3d gfx/wr/swgl/src/vector_type.h
--- a/gfx/wr/swgl/src/vector_type.h	Tue Feb 16 21:17:45 2021 +0000
+++ b/gfx/wr/swgl/src/vector_type.h	Tue Feb 16 21:17:45 2021 +0000
@@ -319,11 +319,20 @@
 #  define yyyy swizzle(1, 1, 1, 1)
 #  define zzzz swizzle(2, 2, 2, 2)
 #  define wwww swizzle(3, 3, 3, 3)
+#  define xxyy swizzle(0, 0, 1, 1)
+#  define xxzz swizzle(0, 0, 2, 2)
+#  define yyww swizzle(1, 1, 3, 3)
+#  define zzww swizzle(2, 2, 3, 3)
 #  define xyxy swizzle(0, 1, 0, 1)
+#  define xzxz swizzle(0, 2, 0, 2)
+#  define ywyw swizzle(1, 3, 1, 3)
 #  define zwzw swizzle(2, 3, 2, 3)
 #  define zwxy swizzle(2, 3, 0, 1)
 #  define zyxw swizzle(2, 1, 0, 3)
+#  define xxyz swizzle(0, 0, 1, 2)
+#  define xyyz swizzle(0, 1, 1, 2)
 #  define xyzz swizzle(0, 1, 2, 2)
+#  define xzyw swizzle(0, 2, 1, 3)
 #  define yzwx swizzle(1, 2, 3, 0)
 #  define wxyz swizzle(3, 0, 1, 2)
 #  define xxxxyyyy XXXXYYYY()
@@ -407,6 +416,16 @@
   return combine(combine(a, b), combine(c, d));
 }
 
+template <typename T, int N>
+SI VectorType<T, N> combineLow(VectorType<T, N> a, VectorType<T, N> b) {
+  return combine(lowHalf(a), lowHalf(b));
+}
+
+template <typename T, int N>
+SI VectorType<T, N> combineHigh(VectorType<T, N> a, VectorType<T, N> b) {
+  return combine(highHalf(a), highHalf(b));
+}
+
 template <typename T>
 SI VectorType<T, 4> zipLow(VectorType<T, 4> a, VectorType<T, 4> b) {
   return SHUFFLE(a, b, 0, 4, 1, 5);
